dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/VisitorInlineAsm.java

```diff
@@ -0,0 +1,619 @@
+            label = this.labelsDefined.get(labelName);
+        } else {
+            label = EventFactory.newLabel(labelName);
+            this.labelsDefined.put(labelName, label);
+        }
+        return label;
+    }
+
+    // This function lets us know which type we need to assign to the created asm register.
+    // In order to do so, we have to understand which llvm register it is going to refer to.
+    // given the registerID of the register e.g. $2 -> registerID = 2
+    // return the type of the llvm register it is referencing
+    // if it is referencing the return register, return its type
+    private Type getLlvmRegisterTypeGivenAsmRegisterID(int registerID) {
+        Type registerType;
+        if (isPartOfReturnRegister(registerID)) {
+            if (returnRegister.getType() instanceof AggregateType at) {
+                // e.g. returnRegisterType = { i32, i32 } "... $1", "=&r, =&r, r" (i32 r5) 
+                //  w.h. registerID = 1, we have to access returnRegister[1].getType()
+                registerType = at.getFields().get(registerID);
+            } else {
+                // returnRegister is not an aggregate, we just get that type
+                registerType = returnRegister.getType();
+            }
+        } else {
+            // in this case we have an index which is higher than the size of the return Register
+            // this means that, like in visitAsmMetadataEntries, we have to "shift" the value by the size of returnRegister type's size
+            // Same example as visitAsmMetadataEntries:
+            // e.g. returnRegisterType = { i32, i32 } "... $2", "=&r, =&r, r" (i32 r5)
+            // we have registerID 2, we have returnRegisterType size = 2, so we have to perform 2-2 = 0, which is the right index
+            // to get argsRegisters[2-2] = argsRegisters[0] = r5.
+            registerType = argsRegisters.get(registerID - getSizeOfReturnRegister()).getType();
+        }
+        return registerType;
+    }
+
+    private String makeRegisterName(int registerID) {
+        return "asm_" + registerID;
+    }
+
+    // this function is the entrypoint of the visitor.
+    // the events that are going to be generated are
+    // 1) inputAssignments -- how we map llvm registers to asm ones
+    // 2) asmInstructions -- the events representing the asm instructions
+    // 3) outputAssignments -- how we map the asm registers to the return Register
+    @Override
+    public List<Event> visitAsm(InlineAsmParser.AsmContext ctx) {
+        // Given this format 
+        //      Register = Type call asm sideeffect 'asm code', 'constraints, clobbers' ('args')
+        // the visitor pattern is going to take care of first visiting all the 'asm code'
+        // which is going to populate all of the 'asmInstructions' and creating the asmRegisters.
+        // Then, by reading the constraints, we are going to be able to understand which asmRegister
+        // maps to which llvmRegister.
```

Comment:
The visitor will first visit the asm code (which will create the events and asm registers) and then the constraints. The latter will take care of creating input and output assignments.

I think the text above is enough and the rest of comments in this method can be removed.

Reply:
> The visitor will first visit the asm code (which will create the events and asm registers) and then the constraints. The latter will take care of creating input and output assignments.

I meant to add this sentence as a description of the overall method in replacement off all the inlined comments